lib/collection/src/grouping/aggregator.rs

```diff
@@ -0,0 +1,425 @@
+use std::collections::{HashMap, HashSet};
+
+use itertools::Itertools;
+use segment::types::{ExtendedPointId, ScoredPoint};
+use serde_json::Value;
+
+use super::types::AggregatorError::{self, *};
+use super::types::{Group, GroupKey, HashablePoint};
+
+type Hits = HashSet<HashablePoint>;
+
+pub(super) struct GroupsAggregator {
+    groups: HashMap<GroupKey, Hits>,
+    max_group_size: usize,
+    grouped_by: String,
+    max_groups: usize,
+    full_groups: HashSet<GroupKey>,
+}
+
+impl GroupsAggregator {
+    pub(super) fn new(groups: usize, group_size: usize, grouped_by: String) -> Self {
+        Self {
+            groups: HashMap::with_capacity(groups),
+            max_group_size: group_size,
+            grouped_by,
+            max_groups: groups,
+            full_groups: HashSet::with_capacity(groups),
+        }
+    }
+
+    /// Adds a point to the group that corresponds based on the group_by field, assumes that the point has the group_by field
+    fn add_point(&mut self, point: &ScoredPoint) -> Result<(), AggregatorError> {
+        if self.full_groups.len() == self.max_groups {
+            return Err(AllGroupsFull);
+        }
+
+        // if the key contains multiple values, grabs the first one
+        let group_key = point
+            .payload
+            .as_ref()
+            .and_then(|p| p.get_value(&self.grouped_by).into_iter().next().cloned())
+            .ok_or(KeyNotFound)
+            .and_then(GroupKey::try_from)?;
+
+        let group = self
+            .groups
+            .entry(group_key.clone())
+            .or_insert_with(|| HashSet::with_capacity(self.max_group_size));
+
+        group.insert(HashablePoint::minimal_from(point));
+
+        if group.len() == self.max_group_size {
+            self.full_groups.insert(group_key);
+        }
+
+        Ok(())
+    }
+
+    /// Adds multiple points to the group that they corresponds based on the group_by field, assumes that the points always have the grouped_by field, else it just ignores them
+    pub(super) fn add_points(&mut self, points: &[ScoredPoint]) {
+        for point in points {
+            if let Err(AllGroupsFull) = self.add_point(point) {
```

Comment:
When matching against enum, I personally prefer exhaustive match for clarity and safety.

Something like

```rust
match self.add_point(point) {
   case Err(AllGroupsFull) => break,
   case Err(KeyNotFound | A | B) => continue,
   case Ok(_) => continue,
}
```